// grunt-tomcat-deploy/tasks/tomcat_deploy.js
/*
 * grunt-tomcat-deploy
 * https://github.com/elebescond/grunt-tomcat-deploy
 *
 * Copyright (c) 2013 Erwan Le Bescond
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {
    grunt.registerTask("tomcat_war",function(){
        var tomcat = grunt.config('tomcat_deploy');
        if(tomcat === undefined){
            return ;
        }
        var archive = tomcat.dist + '.war';
        grunt.loadNpmTasks('grunt-zip');

        grunt.config('zip', {
            war: {
                cwd: tomcat.dist,
                dest: archive,
                src: [tomcat.dist + '/**']
            }
        });
        grunt.task.run('zip');
    });

    grunt.registerTask('tomcat_deploy_only', 'Deploy your files to a tomcat server.', function() {

        var done = this.async();

        grunt.config.requires('tomcat_deploy.login');
        grunt.config.requires('tomcat_deploy.password');
        grunt.config.requires('tomcat_deploy.host');
        grunt.config.requires('tomcat_deploy.port');
        grunt.config.requires('tomcat_deploy.deploy');
        grunt.config.requires('tomcat_deploy.path');

        grunt.task.run(['tomcat_war']);

        var tomcat = grunt.config('tomcat_deploy');

        var archive = tomcat.dist + '.war'

        var options = {
            auth: tomcat.login + ':' + tomcat.password,
            hostname: tomcat.host,
            port: tomcat.port,
            path: tomcat.deploy + '?path=' + tomcat.path + '&update=false',
            method: 'PUT'
        };

        var content = '';

        var req = require('http').request(options, function(res) {
            res.setEncoding('utf8');
            res.on('data', function (chunk){
                console.log("chunk: " + chunk);
                content += chunk;
            });
            res.on('err', function (chunk) {
                grunt.log.error(chunk);
                done(false);
            });
            res.on('end', function (chunk) {
                if(/^OK.*$/m.test(content)) {
                    grunt.log.writeln(content);
                    done();
                }
                else {
                    grunt.log.error(content);
                    done(false);
                }
            });
        });

        var fs = require('fs');

        var stream = fs.createReadStream('./' + archive);

        stream.on('data', function(data) {
            req.write(data);
        });

        stream.on('end', function() {
            req.end();
        });
    });
    var redeploy = function (done ) {
        var tomcat = grunt.config('tomcat_deploy');
        var http = require('http'),
            options = {
                host: tomcat.host,
                port: tomcat.port,
                path: tomcat.path
            };

        var thing = http.get(options, function(res) {
            if( res.statusCode === 404) {
                grunt.log.ok("tomcat_undeploy: Nothing to undeploy");
                grunt.task.run('tomcat_deploy');
                done();
            }
            else {
                grunt.task.run('tomcat_undeploy');
                grunt.task.run('tomcat_deploy');
                done();
            }
        });
    };

    grunt.registerTask('tomcat_deploy', ['tomcat_war', 'tomcat_deploy_only']);

    grunt.registerTask('tomcat_redeploy', function() {
        var async= require('async');
        var gruntDone = this.async();
        async.parallel([
            redeploy
        ], function(){
            gruntDone();
        });
    });



};
